### DISCUSSION TITLE
Merge schemas and renaming commom keys, create custom schemas

### DISCUSSION THREAD

**Discussion by vitormarkis** [colinhacks/zod#2150] [Q&A] [answered]
I'm trying to create a data validation for my database response, it's fetching commentaries from a post. I'll be needing some info from posts schema, users and commentaries.

```tsx
export const userSchema = z.object({
  id: z.number().positive(),
  name: z.string().regex(/ /),
  username: z.string().min(4).max(45),
  email: z.string().email(),
  password: z.string().min(6).max(45),
  profile_pic: z.string().url(),
  cover_pic: z.string().url(),
  city: z.string(),
  website: z.string().url(),
}) 

export const postSchema = z.object({
  id: z.number().positive(),
  text: z.string().max(449),
  image: z.string().url().optional(),
  author_id: z.number().positive(),
  created_at: z.date(),
})

export const commentariesSchema = z.object({
  commentaryId: z.number(),
  text: z.string(),
})
```

There is a lot of cases that those schemas based on database columns, have same column name, such as id, created_at...

How can i create a custom validation merging only the keys i want in this case specifically: 

```tsx
id from commentary as commentaryId
id from posts as postId
text from commentary as com_text: 
created_at from commentary as com_created_at
profile_pic from users as user_profile_pic
name from users as user_name
username from users as user_username
```

I tried to use `.transform` method to rename items, eg: `(post => ({...post, postId: post.id})) ` but it didn't work.

This is verbose but i'm studying the case.

Is this the correct approach to validating database responses?

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const foo = z.object( { id: z.number(), name: z.string() } )
const bar = z.object( { id: z.number(), name: z.string() } )
const schema = z.object( { fooId: foo.shape.id, barId: bar.shape.id } )
console.log( schema.parse( { fooId: 42, barId: 42 } ) ) // { fooId: 42, barId: 42 }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
